x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/persistence/DataFrameTransformsCheckpointService.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.dataframe.persistence;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.stats.IndexShardStats;
+import org.elasticsearch.action.admin.indices.stats.IndexStats;
+import org.elasticsearch.action.admin.indices.stats.IndicesStatsAction;
+import org.elasticsearch.action.admin.indices.stats.IndicesStatsRequest;
+import org.elasticsearch.action.admin.indices.stats.ShardStats;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.dataframe.transforms.DataFrameTransformCheckpoint;
+import org.elasticsearch.xpack.core.dataframe.transforms.DataFrameTransformConfig;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.TreeMap;
+
+/**
+ * DataFrameTransform Checkpoint Service
+ *
+ * Allows checkpointing a source of a data frame transform which includes all relevant checkpoints of the source.
+ *
+ * This will be used to checkpoint a transform, detect changes, run the transform in continuous mode.
+ *
+ */
+public class DataFrameTransformsCheckpointService {
+
+    private static final Logger logger = LogManager.getLogger(DataFrameTransformsCheckpointService.class);
+
+    private final Client client;
+
+    public DataFrameTransformsCheckpointService(final Client client) {
+        this.client = client;
+    }
+
+    /**
+     * Get a checkpoint that is not attached to an id. E.g. for change detection.
+     *
+     * @param transformConfig the @link{DataFrameTransformConfig}
+     * @param listener listener to call after inner request returned
+     */
+    public void getCheckpoint(DataFrameTransformConfig transformConfig, ActionListener<DataFrameTransformCheckpoint> listener) {
+        getCheckpoint(transformConfig, -1L, listener);
+    }
+
+    /**
+     * Get a checkpoint with an id, used to store a checkpoint.
+     *
+     * @param transformConfig the @link{DataFrameTransformConfig}
+     * @param listener listener to call after inner request returned
+     */
+    public void getCheckpoint(DataFrameTransformConfig transformConfig, long checkpointId,
+            ActionListener<DataFrameTransformCheckpoint> listener) {
+        long timestamp = System.currentTimeMillis();
+
+        // placeholder for time based synchronization
+        long timeUpperBound = 0;
+
+        ClientHelper.executeWithHeadersAsync(transformConfig.getHeaders(), ClientHelper.DATA_FRAME_ORIGIN, client,
```

Comment:
I am not sure about requiring the user who created the transform to have access to get indices stats from the source index. It seems like a needless check as anybody who has the ability to read from the transform index will then be able to see the stored `shardStats.getSeqNoStats().getGlobalCheckpoint()`, regardless if they had permissions to read the original source index stats. 

If we are to require permissions to get the indices stats, then we should probably not store the checkpoint information where a user without that permission is able to read it.

Reply:
This raises an interesting question! What if you specify an expression like `all_*` as source but there is `all_cats`, `all_dogs` and `all_the_money` and you do not have access to `all_the_money`. Will the transform validate? What if `all_the_money` has been added after the creation of the transform? I need to test this.

Nevertheless I agree, I think there is no need to call this with headers.

Reply:
I am now checking the indexes with the user right (requires metadata access rights) but getting the stats without using user headers. The output is filtered according to the indexes the user has access to.